### PR TITLE
feat!: account for time already elapsed when waiting after the commit

### PR DESCRIPTION
**Pull request by evan-forbes**
## Description

This PR uses dynamic timeout commits, meaning that we account for time that has already elapsed before we decide how long to wait after the commit is finalized. This effectively adds a "target round duration", where we don't wait the same amount of time for the timeout commit arbitrarily each round. Instead, we subtract the time elapsed between the round's start time, and the commit time. 

Note that this will likely have yet unmeasured consequences in the application. For instance, if we somehow end up not waiting at all for remaining votes, then those votes won't get included in the commit, and validators that otherwise would have gotten counted as signing (and thus gotten rewards), will not be counted.

part of #939 and https://github.com/celestiaorg/celestia-app/issues/1340

### CODE REVIEW

**Comment by evan-forbes** [comment]
marking ready for review to begin accepting comments. while this change is small and seems to be working here and in the app, we should expect other consequences.

this has also only been tested in a contrived testnets, we should ideally test this in testground before including in arabica and mocha

File: config/config.go
Comment (rootulp) [inline config/config.go:934]: [question] I'm confused on the relationship between `TimeoutCommit`, `TargetRoundDuration` and `time_iota_ms`. Does `time_iota_ms` still need to be updated after this PR? If not,

```suggestion
```

File: config/config.go
Comment (rootulp) [inline config/config.go:980]: same question as above. If this comment is no longer applicable:

```suggestion
```

File: config/config.go
Comment (evan-forbes) [inline config/config.go:961]: I'm open to better names for this

File: config/config.go
Comment (rootulp) [inline config/config.go:962]: [for reviewers] noting `SkipTimeoutCommit=false` is already the case. Ref: https://github.com/tendermint/tendermint/issues/5911#issuecomment-967615517

File: config/config.go
Comment (rootulp) [inline config/config.go:961]: [question] is there a guaranteed # of rounds per block? I assume no, so does this change mean each round will occur approx. every 3.5 seconds and correspondingly a block will occur approx. every 3.5 seconds * number of rounds in a block?

Comment (rootulp) [reply config/config.go:961]: [question] on a related note, do we have data on the # of rounds per block for our current testnets? Is that how the 3.5 seconds was arrived at? Potentially 15 second target block time ~= 4 rounds * 3.5 seconds (?)

File: config/config.go
Comment (rootulp) [inline config/config.go:963]: [for reviewers] noting that `CreateEmptyBlocks=true` is already the case and this seems desirable if we want consistent block times per https://docs.tendermint.com/v0.34/tendermint-core/configuration.html#create-empty-blocks-true

File: config/config.go
Comment (rootulp) [inline config/config.go:964]: [question] should this be 15 seconds given that's our desired block time?

```suggestion
		CreateEmptyBlocksInterval:   15 * time.Second,
```

**Review by rootulp — COMMENTED** [review]
No blocking feedback. IIUC it isn't possible for tendermint to guarantee a consistent block interval but this change tries to target a more consistent block time.

Relevant context: https://github.com/tendermint/tendermint/issues/5911 and https://docs.tendermint.com/v0.34/tendermint-core/configuration.html#consensus-timeouts-explained

File: config/config.go
Comment (evan-forbes) [reply config/config.go:964]: see above comment, we should set this in the app for a few different reasons. One is simply to avoid fixing all the tests that break and causing a huge diff

Comment (evan-forbes) [reply config/config.go:964]: imo no, at least not atm, as that is an optimization to avoid empty, and complicates timings. instead, I think we should focus on finding a solution that works for all blocks, such as setting the TargetRoundDuration to 15s (in the app ofc). If a tx appears in the mempool in that period of time, then it will simply be included in the next block

File: config/config.go
Comment (evan-forbes) [reply config/config.go:934]: In theory, time iota should do what we want by setting a consensus param for the minimum time between commits, but in practice, setting the value to something larger than the actual block time messes with all of the timings and breaks header verification. Specifically when we check the times of the headers, the times often creep into the future. I think this comment is either outdated or is still relevant is the block times are shorter than the time_iota value

iirc, time iota is being removed in the next version of tendermint. Regardless though, I'm typically in favor of not changing comments like this in this repo unless truly necessary, and instead addressing them upstream if they need to be changed.

File: config/config.go
Comment (liamsi) [inline config/config.go:1020]: This looks like a reasonable, simple change to me. Would it make sense to:
 - test this in a live network or at least in testground
 - have informal review this as well
before we merge this?

File: config/config.go
Comment (rootulp) [reply config/config.go:934]: > iirc, time iota is being removed in the next version of tendermint. Regardless though, I'm typically in favor of not changing comments like this in this repo unless truly necessary, and instead addressing them upstream if they need to be changed.

Agreed, I didn't realize this comment also exists in vanilla tendermint/tendermint [here](https://github.com/tendermint/tendermint/blob/35581cf54ec436b8c37fabb43fdaa3f48339a170/config/config.go#L937).

**Review by rootulp — DISMISSED** [review]

File: config/config.go
Comment (evan-forbes) [reply config/config.go:1020]: yes that's a good idea. We were trying to test this with testground this morning, but didn't have time to finish. 

We might not want to include this change in the incentivized testnet, but we should push and test in mocha as soon as we can get data back from testground.

One side thought: after playing around with this, I noticed that if we increase the timeouts too long, then validators begin to miss producing blocks, even on the e2e test. After discussing with @cmwaters, when the timeouts are long, we think there is a significantly higher probability validators are progressing through consensus at different rates, and thus potentially missing rounds entirely. This would actually explain what we are seeing in mocha quite well. Having more consistent round times, potentially via this mechansim, but likely via similar arbitary waiting mechanisms to more closely follow the configured timeouts instead of moving to the next step in consensus after reaching 2/3s, could be effective in helping us come to consensus on the first block when we expect.

File: config/config.go
Comment (cmwaters) [reply config/config.go:964]: I believe `CreateEmptyBlocksInterval` refers to how long to wait without any transactions in the mempool before going ahead and forming that empty block. Which means if transactions arrive, the block will be constructed far sooner than 15 seconds and it also does not consider how long it takes for consensus to be reached on that block

File: config/config.go
Comment (cmwaters) [inline config/config.go:935]: It's not really per round but per height because we only commit once per height. What this means of course is that any multiround height will likely be above 15 seconds and instantly jump to the next height meaning only 2/3 voting power will be included in the commit (i.e. there will be no time for straggler voters). We should be wary of this if we are rewarding validators on a per-height basis for voting involvement. Alternatively we could reward validators for involvement across 100 heights (instead of just one). Or we could introduce another parameter which states a minimum "wait for straggler votes" period. 

Actually, It might be fine if there's also a non-negligible time taken in `FinalizeBlock` and `PrepareProposal`. If it takes 1-2 seconds, then that's another 1 to 2 seconds that more votes (above the 2/3 line) can be received in. 

We also might not care to have commits with all signatures.

Conclusion: I think we should rename this variable to `TargetHeightDuration` and we should make sure we are measuring the percentage of voting power in commits.

File: config/config.go
Comment (cmwaters) [reply config/config.go:962]: I think we should force this to be `false` (in app) else it will mess with this logic

File: config/toml.go
Comment (cmwaters) [inline config/toml.go:463]: Maybe extend the comments here

File: consensus/state.go
Comment (cmwaters) [inline consensus/state.go:658]: For me, this is more complex than it needs to be. Can't you just have:
```suggestion
		cs.StartTime = cs.config.Commit(cs.StartTime)
```
and
```go
func (cfg *ConsensusConfig) Commit(startTime time.Time) {
    return startTime.Add(cfg.TargetRoundDuration)
}
```
It seems like consensus can handle `StartTime`'s in the past so I don't think you need to worrry about it being less than now. `CommitTime` really has little to do with it. i.e. 
CommitTime + (TargetRoundDuration - (CommitTime-StartTime) = StartTime + TargetRoundDuration.

Comment (cmwaters) [reply consensus/state.go:658]: I'd actually rename `Commit(...)` to `NextStartTime(...)`. It seems more appropriate

**Review by cmwaters — CHANGES_REQUESTED** [review]
I've tried to think through this a bit and I think it all makes sense. If you really wanted regular block times, we should implement proposer based timestamps, but this should get us something in the range. We already have metrics for block times right? So would be good to track them.

**Comment by evan-forbes** [comment]
I still need to debug why the e2e test is failing after using a simpler timeout, but we also shouldn't merge this in until we collect more data on this change's impact via testground, so I'm converting to a draft

File: config/config.go
Comment (evan-forbes) [reply config/config.go]: > It's not really per round but per height because we only commit once per height.

ahh I see! good catch. I wrongly assumed that since the start time is in `RoundState` it would get updated each round. Renamed as well

> We should be wary of this if we are rewarding validators on a per-height basis for voting involvement.

I was worried about this as well (briefly mentioned in top comment) as iirc, there was somewhere in the sdk where we're rewarding validators per signature. I looked again though and couldn't find anything, so it might be obfuscated. Regardless, the number of signatures included should be something that we collect data on in testground, and might even want to force this to occur if it isn't happening naturally.

File: consensus/state.go
Comment (evan-forbes) [reply consensus/state.go]: good idea, renamed

File: DOCKER/docker-entrypoint.sh
Comment (rootulp) [inline DOCKER/docker-entrypoint.sh:12]: `target_round_duration` seems like the previous name for `target_height_duration`

```suggestion
		-e 's/^target_height_duration\s*=.*/target_height_duration = "1000ms"/' \
```

File: config/toml.go
Comment (rootulp) [inline config/toml.go:464]:
```suggestion
target_height_duration = "{{ .Consensus.TargetHeightDuration }}"
```

File: docs/tendermint-core/configuration.md
Comment (rootulp) [inline docs/tendermint-core/configuration.md:460]: [optional] these are just docs so the value doesn't matter but proposal for it to match the 15s on line 354

```suggestion
target_height_duration = "15s"
```

File: config/config.go
Comment (rootulp) [inline config/config.go:960]: [question] why isn't this 15 seconds?

```suggestion
		TargetHeightDuration:        15 * time.Second,
```

Comment (evan-forbes) [reply config/config.go:960]: it will break a bunch of tests, since they're testing assuming specific block times

File: config/toml.go
Comment (evan-forbes) [reply config/toml.go:464]: https://github.com/celestiaorg/celestia-core/pull/965/commits/8e5b17b05ac3f4c03d9ae7c9eed0155cd49ebaff

File: DOCKER/docker-entrypoint.sh
Comment (evan-forbes) [reply DOCKER/docker-entrypoint.sh]: https://github.com/celestiaorg/celestia-core/pull/965/commits/8e5b17b05ac3f4c03d9ae7c9eed0155cd49ebaff

File: config/config.go
Comment (rootulp) [reply config/config.go:960]: got it, no change needed. Resolving.

File: consensus/state.go
Comment (evan-forbes) [inline consensus/state.go:682]: I'll abstract this out and unit test it, then we still need to run testground tests, along with an overally analysis of selecting ideal timings, but this is what is required to get consistent block times in the e2e test. (I'll aslo remove the event collector)

File: consensus/state.go
Comment (evan-forbes) [inline consensus/state.go:674]: @cmwaters do you know why this elapsed value can sometimes be negative?

**Comment by cmwaters** [comment]
I think it's okay if we go ahead and merge this. This version won't be used in v1 but in v2

File: consensus/state.go
Comment (cmwaters) [inline consensus/state.go:705]: Do we still want to add this?

File: test/maverick/consensus/state.go
Comment (cmwaters) [inline test/maverick/consensus/state.go:910]:
```suggestion
                cs.StartTime = cs.config.NextStartTime(cs.StartTime)
```

File: test/maverick/consensus/state.go
Comment (cmwaters) [inline test/maverick/consensus/state.go:903]:
```suggestion
		cs.StartTime = state.LastBlockTime
```

**Review by cmwaters — APPROVED** [review]
LGTM

**Comment by evan-forbes** [comment]
forgot to update this with tests that were run on testground also had 14-16s block times, without having any heights that required more than a single round of consensus, so I think we're clear to try this on a testnet

**Review by rootulp — APPROVED** [review]

**Review by cmwaters — APPROVED** [review]
LGTM

**Pull request merged**
